Document Non-Class-Based Testing (#35944)
gradle/gradle: platforms/documentation/docs/src/docs/release/notes.md

See the [Gradle 9.x upgrade guide](userguide/upgrading_version_9.html#changes_@baseVersion@) to learn about deprecations, breaking changes, and other considerations when upgrading to Gradle @version@.

For Java, Groovy, Kotlin, and Android compatibility, see the [full compatibility notes](userguide/compatibility.html).

## New features and usability improvements

<!-- Do not add breaking changes or deprecations here! Add them to the upgrade guide instead. -->

### Non-Class Based Testing

> When testing using JUnit Platform, Gradle can now discover and execute tests that are not defined in classes.

Reviewer: Is it possible to add a bit more about the motivation for this change here:
```suggestion
When using JUnit Platform, Gradle can now discover and execute tests that are not defined in JVM classes.
```

```kotlin
    testing.suites.named("test", JvmTestSuite::class) {
        useJUnitJupiter()

        dependencies {
            implementation("io.cucumber:cucumber-java:7.15.0")
            runtimeOnly("io.cucumber:cucumber-junit-platform-engine:7.15.0")
        }

        targets.all {
            testTask.configure {
                testDefinitionDirs.from("src/test/resources")  // Conventional Cucumber *.feature files location
            }
        }
    }
```

### Additional test data capture

During test execution JUnit Platform tests can emit additional data such as file attachments or arbitrary key-value pairs using it's [TestReporter API](https://docs.junit.org/current/user-guide/#writing-tests-dependency-injection).

This typically looks something like:

```java
    @Test
    void someTestMethod(TestReporter testReporter) {
        testReporter.publishEntry("myKey", "myValue");
        testReporter.publishFile("test1.txt", MediaType.TEXT_PLAIN_UTF_8, file -> Files.write(file, List.of("Test 1")));
        // Test logic continues...
    }
```

Gradle now captures this additional data and makes it available in HTML test reports and XML test results files.

> The HTML test report presents the data in 2 new `data` and `attachement` tabs, that will appear similarly to the existing `stdout` and `stderr` tabs, when this data is present.

Reviewer: ```suggestion
In the HTML test report, when such `data` is present, two new tabs are shown alongside `stdout` and `stderr`:

- Data – for key–value entries
- Attachments – for file attachments
```

The JUnit XML report presents the data in the following manner:

> - `ReportEntry` are present as `<properties/>`

Reviewer: ```suggestion
- `ReportEntry` values as `<properties/>`
```

> - `FileEntry` are present as `[[ATTACHMENT|/path/to/file]]` based on conventions used by Jenkins, Azure pipelines and GitLab.

Reviewer: ```suggestion
- `FileEntry` values as `[[ATTACHMENT|/path/to/file]]`, following conventions used by Jenkins, Azure Pipelines, and GitLab
```

vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv -->

### Non-Class Based Testing

> When testing using JUnit Platform, Gradle can now discover and execute tests that are not defined in classes.

Reviewer: ```suggestion
When testing using [JUnit Platform](https://junit.org/), Gradle can now discover and execute tests that are not defined in classes.
```

JUnit Platform [TestEngine](https://docs.junit.org/current/user-guide/#test-engines)s are able to discover and execute tests defined in arbitrary formats, and are not limited to JVM classes.
Previously, however, Gradle's [Test](dsl/org.gradle.api.tasks.testing.Test.html) task required test classes to be present; otherwise execution would fail with a message:

```text
There are test sources present and no filters are applied, but the test task did not discover any tests to execute.
```

> Now, tests can be defined in whatever format is understood by the configured `TestEngine`s.

Reviewer: Does it need to be plural?
```suggestion
In this release, tests can be defined in whatever format is understood by the configured `TestEngine`.
```

These non-class-based tests can now be run directly without workarounds:

```kotlin
    testing.suites.named("test", JvmTestSuite::class) {
        useJUnitJupiter()

        dependencies {
            implementation("io.cucumber:cucumber-java:7.15.0")
            runtimeOnly("io.cucumber:cucumber-junit-platform-engine:7.15.0")
        }

        targets.all {
            testTask.configure {
                testDefinitionDirs.from("src/test/resources")  // Conventional Cucumber *.feature files location
            }
        }
    }
```

### Additional test data capture

During test execution, JUnit Platform tests can emit additional data such as file attachments or arbitrary key–value pairs using its [TestReporter API](https://docs.junit.org/current/user-guide/#writing-tests-dependency-injection).

For example:

```java
    @Test
    void someTestMethod(TestReporter testReporter) {
        testReporter.publishEntry("myKey", "myValue");
        testReporter.publishFile("test1.txt", MediaType.TEXT_PLAIN_UTF_8, file -> Files.write(file, List.of("Test 1")));
        // Test logic continues...
    }
```

> Gradle now captures this additional data and includes it in both the HTML test report and the XML test results.

Reviewer: Let's link to the docs about the HTML test report here.

vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv -->

> ### Non-Class Based Testing

Author: Yeah, something weird happened here.  I've cleaned it up.
